src/Clapi/Types/SequenceOps.hs

```diff
@@ -0,0 +1,79 @@
+{-# OPTIONS_GHC -Wall -Wno-orphans #-}
+
+module Clapi.Types.SequenceOps
+  ( ReorderBundle(..), SequenceOp(..)
+  , digest, produce, applyDigest
+  ) where
+
+import Prelude hiding (fail)
+import Control.Monad.Fail (MonadFail(..))
+import qualified Data.Map as Map
+import qualified Data.List as List
+import Data.Foldable (foldlM)
+
+import Clapi.Types.UniqList (UniqList, unUniqList, mkUniqList)
+
+data SequenceOp i
+  = AddAfter (Maybe i)
+  | MoveAfter (Maybe i)
+  | DelElem
+  deriving (Eq, Show)
+
+newtype ReorderBundle i = ReorderBundle [(i, SequenceOp i)] deriving (Eq, Show)
+newtype SequenceDigest i = SequenceDigest (Map.Map i (SequenceOp i))
+
+sdEmpty :: SequenceDigest i -> Bool
+sdEmpty (SequenceDigest m) = null m
+
+digest :: Ord i => ReorderBundle i -> SequenceDigest i
+digest (ReorderBundle ops) = SequenceDigest $ Map.fromList ops
+
+produce :: SequenceDigest i -> ReorderBundle i
+produce (SequenceDigest m) = ReorderBundle $ Map.toList m
+
+getChainStarts ::
+    Ord i => SequenceDigest i -> ([(i, SequenceOp i)], SequenceDigest i)
+getChainStarts (SequenceDigest m) =
+    let
+        getDep (AddAfter mi) = mi
+        getDep (MoveAfter mi) = mi
+        getDep DelElem = Nothing
+        hasUnresolvedDep = maybe False (flip Map.member m) . getDep
+        (remainder, starts) = Map.partition hasUnresolvedDep m
+    in (Map.toList starts, SequenceDigest remainder)
+
+applyDigest
+    :: (MonadFail m, Ord i, Show i)
+    => SequenceDigest i -> UniqList i -> m (UniqList i)
+applyDigest sd ul =
+    resolveDigest sd >>= applyOps (unUniqList ul) >>= mkUniqList
+  where
+    resolveDigest sd' = if sdEmpty sd' then return []
+        else case getChainStarts sd' of
+            ([], _) -> fail "Unresolvable order dependencies"
+            (starts, remainder) -> (starts ++) <$> resolveDigest remainder
+    applyOps l starts = foldlM applyOp l starts
+
+applyOp :: (MonadFail m, Ord i, Show i) => [i] -> (i, SequenceOp i) -> m [i]
+applyOp l (i, op) = case op of
+    AddAfter mi -> if i `elem` l
+        then fail "Cannot add element twice"
+        else insertAfter i mi l
+    MoveAfter mi ->
+        removeElem "Element was not present to move" i l >>= insertAfter i mi
+    DelElem -> removeElem "Element not present to remove" i l
+  where
+    insertAfter v mAfter ol = case mAfter of
+        Nothing -> return $ v : ol
+        Just after ->
+          let
+            (bl, al) = span (/= after) ol
+          in case al of
+            (a:rl) -> return $ bl ++ [a, v] ++ rl
+            [] -> fail "Cannot add after non-existant element"
```

Comment:
This leads to the message about "adding" if I've done a move, which seems a little odd. (Highlighted in your test on line 58.) Perhaps do the same "pass me the message" thing as you did with `removeElem`?